packages/stream-receiver/src/index.ts

```diff
@@ -0,0 +1,430 @@
+import { isValidAssetScale } from 'ilp-protocol-ildcp'
+import { Writer, Reader } from 'oer-utils'
+import createLogger from 'ilp-logger'
+import {
+  IlpReject,
+  IlpPrepare,
+  IlpFulfill,
+  IlpPacketType,
+  IlpAddress,
+  isValidIlpAddress,
+  IlpError,
+  IlpReply,
+} from 'ilp-packet'
+import { hmac, sha256, base64url, ReplyBuilder, decrypt, encrypt } from './utils'
+import {
+  Packet,
+  ConnectionCloseFrame,
+  FrameType,
+  ErrorCode,
+  ConnectionAssetDetailsFrame,
+  StreamReceiptFrame,
+} from 'ilp-protocol-stream/dist/src/packet'
+import { LongValue } from 'ilp-protocol-stream/dist/src/util/long'
+import { createReceipt } from 'ilp-protocol-stream/dist/src/util/receipt'
+import Long from 'long'
+
+/** Parameters to statelessly generate new STREAM connections and handle incoming packets for STREAM connections. */
+export interface ServerOptions {
+  /** Secret used to stateless generate credentials for incoming STREAM connections. */
+  serverSecret: Buffer
+
+  /** Base ILP address of this STREAM server to access it over the Interledger network. */
+  serverAddress: string
+}
+
+/** Credentials for a client to setup a STREAM connection with a STREAM server */
+export interface StreamCredentials {
+  /** ILP address of the recipient account, identifying this connection, for the client to send packets to this STREAM server. */
+  ilpAddress: IlpAddress
+
+  /** 32-byte seed to encrypt and decrypt STREAM messages, and generate ILP packet fulfillments. */
+  sharedSecret: Buffer
+}
+
+/** Pending STREAM request and in-flight ILP Prepare with funds that may be fulfilled or rejected. */
+export interface IncomingMoney {
+  /** Unique identifier of this STREAM connection: SHA-256 hash of destination ILP address with token, hex-encoded */
+  connectionId: string
+
+  /** Arbitrary data to attribute or handle an incoming payment, encoded when the credentials were generated. */
+  paymentTag?: string
+
+  /**
+   * Sign and include and STREAM receipt for the total amount received on this STREAM connection, per `connectionId`,
+   * including the additional amount from this packet. Amount must be within the u64 range.
+   */
+  setTotalReceived(totalReceived: LongValue): void
+
+  /** Fulfill the money from this incoming ILP Prepare packet */
+  accept(): IlpFulfill
+
+  /** Temporarily decline the incoming money: inform STREAM sender to backoff in time (T00 Reject: Temporary Internal Error) */
+  softDecline(): IlpReject
```

Comment:
What do you think of `temporaryDecline` and `finalDecline` to more closely mirror the ILP error names? (I don't feel strongly one way or the other -- I'm not sure which will be more intuitive).

Reply:
https://github.com/interledgerjs/interledgerjs/compare/b6a4203...ko/stream-receiver#diff-e79c6bb9baee213b784c3248393f0ed8R296-R297